### PR TITLE
E UI 4878 hearing generic selection

### PR DESCRIPTION
**Pull request by waynepHMCTS**
### JIRA link (if applicable) ###

https://tools.hmcts.net/jira/browse/EUI-4878

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[x ] No
```

### CODE REVIEW

File: src/hearings/components/multi-level-selector/multi-level-selector.component.spec.ts [hunk 1]
```diff
@@ -0,0 +1,216 @@
...
+      },
+      {
+        controlType: ControlTypeEnum.SELECT,
+        level: 2,
+      }
+    ];
+    fixture.detectChanges();
+  });
+
+  it('should create', () => {
```
Comment (kuda1992) [inline src/hearings/components/multi-level-selector/multi-level-selector.component.spec.ts:212]: Are we actually testing the component 😭.

File: src/hearings/containers/request-hearing/hearing-panel/hearing-panel.component.ts [hunk 2]
```diff
@@ -1,42 +1,124 @@
...
   public initForm(): void {
     this.panelJudgeForm = this.formBuilder.group({
       specificPanel: ['', Validators.required],
+      multiSelect: this.formBuilder.array([], this.arrayValidator(1))
+    });
+
+    this.panelJudgeForm.controls.multiSelect = this.convertRefDataModelToArray(this.multiSelection);
+  }
+
+  public convertArrayToRefDataMode(array: FormArray): RefDataModel[] {
```
Comment (kuda1992) [inline src/hearings/containers/request-hearing/hearing-panel/hearing-panel.component.ts:74]: Minor spelling mistake. Should be `convertArrayToRefDataModel ` not `convertArrayToRefDataMode`

File: src/hearings/models/hearings.enum.ts [hunk 3]
```diff
@@ -206,3 +207,11 @@ export enum Mode {
   VIEW = 'view',
   VIEW_EDIT = 'view-edit'
 }
+
+export enum ControlTypeEnum {
+  TEXT_BOX,
```
Comment (kuda1992) [inline src/hearings/models/hearings.enum.ts:212]: According to the [typescript guidelines](https://www.typescriptlang.org/docs/handbook/enums.html) enums should be `PascalCase` not `UPPER_CASE_SNAKE_CASE `. Minor though.

File: api/hearings/models/refData.model.ts [hunk 4]
```diff
@@ -6,6 +6,7 @@ export interface RefDataModel {
   hintTextCY: string;
   order: number;
   parentKey: string;
+  child_nodes?: RefDataModel[];
```
Comment (kuda1992) [inline api/hearings/models/refData.model.ts:9]: Does the data from the api come back as `snake_case`? If not can we stick to `camelCase`?

**Review by kuda1992 — CHANGES_REQUESTED** [review]
Very minor nitpick changes, but mostly good to me. 😊

File: src/hearings/components/multi-level-selector/multi-level-selector.component.spec.ts (hunk 1 above)
Comment (waynepHMCTS) [reply src/hearings/components/multi-level-selector/multi-level-selector.component.spec.ts:212]: Unit test result is ok and Code coverage is 95%. So yes. Are you expecting more tests?

File: src/hearings/models/hearings.enum.ts (hunk 3 above)
Comment (waynepHMCTS) [reply src/hearings/models/hearings.enum.ts:212]: Hi @kuda1992 and @rayliangatos  the hearings.enum has a mixture off camalCase and PascalCase could you guys decide decide which one.

File: api/hearings/models/refData.model.ts (hunk 4 above)
Comment (waynepHMCTS) [reply api/hearings/models/refData.model.ts:9]: I think we need a quick chat about coding standards @rayliangatos and @kuda1992

File: src/hearings/models/hearings.enum.ts (hunk 3 above)
Comment (rayliangatos) [reply src/hearings/models/hearings.enum.ts:212]: @kuda1992 where is the guideline for using PascalCase in typescript as enum? I couldn't find it on the link you provided, as it only has a single word in its example.

File: api/hearings/models/refData.model.ts (hunk 4 above)
Comment (rayliangatos) [reply api/hearings/models/refData.model.ts:9]: yes this child_nodes is from api,

Comment (kuda1992) [reply api/hearings/models/refData.model.ts:9]: @rayliangatos @waynepHMCTS . If that's the case I will resolve the conversation.

File: src/hearings/components/multi-level-selector/multi-level-selector.component.spec.ts (hunk 1 above)
Comment (kuda1992) [reply src/hearings/components/multi-level-selector/multi-level-selector.component.spec.ts:212]: Kinder maybe we can test elements rendered, given inputs provided.

File: api/hearings/data/reference.mock.data.ts [hunk 5]
```diff
@@ -240,6 +243,131 @@ export const ADDITIONAL_HEARING_FACILITIES_REF: RefDataModel[] = [
...
 ];
+
+export const LIST_OFF_VALUES_REF: RefDataModel[] = [{
+  key: 'DisabilityQualifiedPanelMember',
+  value_en: 'Disability qualified panel member',
+  value_cy: '',
+  hintText_EN: 'true',
+  hintTextCY: 'false',
+  order: 1,
+  parentKey: '3',
```
Comment (rayliangatos) [inline api/hearings/data/reference.mock.data.ts:254]: Though this is not used but parentKey should be null if no parent.

File: api/hearings/data/reference.mock.data.ts [hunk 6]
```diff
@@ -240,6 +243,131 @@ export const ADDITIONAL_HEARING_FACILITIES_REF: RefDataModel[] = [
...
+}, {
+  key: 'MedicallyQualifiedPanelMember1',
+  value_en: 'Medically qualified panel member',
+  value_cy: '',
+  hintText_EN: 'true',
+  hintTextCY: 'false',
+  order: 1,
+  parentKey: '3',
```
Comment (rayliangatos) [inline api/hearings/data/reference.mock.data.ts:262]: Though this is not used but parentKey should be null if no parent.

File: api/hearings/data/reference.mock.data.ts [hunk 7]
```diff
@@ -240,6 +243,131 @@ export const ADDITIONAL_HEARING_FACILITIES_REF: RefDataModel[] = [
...
+  child_nodes: [{
+    key: 'Cardiologist',
+    value_en: 'Cardiologist',
+    value_cy: '',
+    hintText_EN: 'true',
+    hintTextCY: 'false',
+    order: 1,
+    parentKey: '3',
```
Comment (rayliangatos) [inline api/hearings/data/reference.mock.data.ts:270]: parentKey should be 'MedicallyQualifiedPanelMember1'

File: api/hearings/data/reference.mock.data.ts [hunk 8]
```diff
@@ -240,6 +243,131 @@ export const ADDITIONAL_HEARING_FACILITIES_REF: RefDataModel[] = [
...
+    child_nodes: [],
+  },
+  {
+    key: 'Carer',
+    value_en: 'Carer',
+    value_cy: '',
+    hintText_EN: 'true',
+    hintTextCY: 'false',
+    order: 1,
+    parentKey: '3',
```
Comment (rayliangatos) [inline api/hearings/data/reference.mock.data.ts:280]: ditto

File: api/hearings/data/reference.mock.data.ts [hunk 9]
```diff
@@ -240,6 +243,131 @@ export const ADDITIONAL_HEARING_FACILITIES_REF: RefDataModel[] = [
...
+    child_nodes: [],
+  },
+  {
+    key: 'EyeSurgeon',
+    value_en: 'Medically qualified panel member',
+    value_cy: '',
+    hintText_EN: 'true',
+    hintTextCY: 'false',
+    order: 1,
+    parentKey: '3',
```
Comment (rayliangatos) [inline api/hearings/data/reference.mock.data.ts:290]: ditto

File: api/hearings/data/reference.mock.data.ts [hunk 10]
```diff
@@ -240,6 +243,131 @@ export const ADDITIONAL_HEARING_FACILITIES_REF: RefDataModel[] = [
...
+    child_nodes: [],
+  },
+  {
+    key: 'GeneralPractitioner',
+    value_en: 'General Practitioner',
+    value_cy: '',
+    hintText_EN: 'true',
+    hintTextCY: 'false',
+    order: 1,
+    parentKey: '3',
```
Comment (rayliangatos) [inline api/hearings/data/reference.mock.data.ts:300]: ditto

File: api/hearings/data/reference.mock.data.ts [hunk 11]
```diff
@@ -240,6 +243,131 @@ export const ADDITIONAL_HEARING_FACILITIES_REF: RefDataModel[] = [
...
+    child_nodes: [],
+  }],
+}, {
+  key: 'MedicallyQualifiedPanelMember',
+  value_en: 'Medically qualified panel member',
+  value_cy: '',
+  hintText_EN: 'true',
+  hintTextCY: 'false',
+  order: 1,
+  parentKey: '3',
```
Comment (rayliangatos) [inline api/hearings/data/reference.mock.data.ts:310]: Though this is not used but parentKey should be null if no parent.

File: src/hearings/models/hearings.enum.ts [hunk 12]
```diff
@@ -97,6 +97,7 @@ export enum HearingCategory {
   NonStdDurationReasonCodes = 'NonStdDurationReasonCodes',
   EntityRoleCodes = 'EntityRoleCodes',
   JudgeType = 'JudgeType',
+  ListOffValues = 'ListOffValues',
```
Comment (rayliangatos) [inline src/hearings/models/hearings.enum.ts:100]: This LoV is a generic method, it should be category specific, so I think it should be something like 'OtherPanelRoles'

File: src/hearings/containers/request-hearing/hearing-panel/hearing-panel.component.ts [hunk 13]
```diff
@@ -58,10 +140,11 @@ export class HearingPanelComponent extends RequestHearingPageFlow implements OnI
 
   public checkFormData(): void {
     this.validationErrors = [];
+    this.convertArrayToRefDataMode(this.panelJudgeForm.controls.multiSelect as FormArray);
```
Comment (rayliangatos) [inline src/hearings/containers/request-hearing/hearing-panel/hearing-panel.component.ts]: Mode->Model

File: src/hearings/containers/request-hearing/hearing-panel/hearing-panel.component.ts [hunk 14]
```diff
@@ -1,42 +1,124 @@
...
-import {Store} from '@ngrx/store';
-import {HearingJudgeNamesListComponent} from '../../../components';
-import {ACTION, HearingPanelSelectionEnum} from '../../../models/hearings.enum';
-import {HearingsService} from '../../../services/hearings.service';
+import { AfterViewInit, Component, OnDestroy, OnInit, ViewChild } from '@angular/core';
+import { AbstractControl, FormArray, FormBuilder, FormGroup, Validators } from '@angular/forms';
+import { ActivatedRoute } from '@angular/router';
+import { Person } from '@hmcts/rpx-xui-common-lib';
+import { Store } from '@ngrx/store';
+import { ControlTypeEnum } from 'src/hearings/models/hearings.enum';
```
Comment (rayliangatos) [inline src/hearings/containers/request-hearing/hearing-panel/hearing-panel.component.ts:6]: src

File: src/hearings/components/multi-level-selector/multi-level-selector.component.ts [hunk 15]
```diff
@@ -0,0 +1,35 @@
+import { Component, Input } from '@angular/core';
+import { FormArray, FormBuilder, FormGroup } from '@angular/forms';
+import { ControlTypeEnum } from 'src/hearings/models/hearings.enum';
+
+@Component({
+  selector: 'exui-multi-level-selector',
+  templateUrl: './multi-level-selector.component.html',
+  styles: ['.multi-select-container { margin-top: 10px; }']
```
Comment (rayliangatos) [inline src/hearings/components/multi-level-selector/multi-level-selector.component.ts:8]: can this refer to be sscs file?

File: api/hearings/data/reference.mock.data.ts [hunk 16]
```diff
@@ -266,6 +394,10 @@ export const ADDITIONAL_HEARING_FACILITIES: RefDataByServiceModel = {
   serviceID: 'SSCS',
   values: ADDITIONAL_HEARING_FACILITIES_REF,
 };
+export const LIST_OFF_VALUES: RefDataByServiceModel = {
```
Comment (rayliangatos) [inline api/hearings/data/reference.mock.data.ts:397]: As this is a generic API LoV(List Of Values), this LIST_OFF_VALUES is too generic here, it should be something specific to SSCS_OTHER_PANEL_ROLES

File: src/hearings/components/multi-level-selector/multi-level-selector.component.ts [hunk 17]
```diff
@@ -0,0 +1,35 @@
...
+})
+export class MultiLevelSelectorComponent {
+  @Input() public multiSelect: FormArray;
```
Comment (rayliangatos) [inline src/hearings/components/multi-level-selector/multi-level-selector.component.ts:11]: if this is also supporting radio, would this still call multiSelect?

File: src/hearings/components/multi-level-selector/multi-level-selector.component.ts [hunk 18]
```diff
@@ -0,0 +1,35 @@
+import { Component, Input } from '@angular/core';
+import { FormArray, FormBuilder, FormGroup } from '@angular/forms';
+import { ControlTypeEnum } from 'src/hearings/models/hearings.enum';
```
Comment (rayliangatos) [inline src/hearings/components/multi-level-selector/multi-level-selector.component.ts:3]: src

File: src/hearings/components/multi-level-selector/multi-level-selector.component.spec.ts [hunk 19]
```diff
@@ -0,0 +1,216 @@
+import { ComponentFixture, TestBed } from '@angular/core/testing';
+import { FormArray, FormBuilder, FormGroup, ReactiveFormsModule, Validators } from '@angular/forms';
+import { ControlTypeEnum } from 'src/hearings/models/hearings.enum';
+import { RefDataModel } from 'src/hearings/models/refData.model';
```
Comment (rayliangatos) [inline src/hearings/components/multi-level-selector/multi-level-selector.component.spec.ts:4]: srcs

File: src/hearings/components/multi-level-selector/multi-level-selector.component.html [hunk 20]
```diff
@@ -0,0 +1,46 @@
...
+        </exui-multi-level-selector>
+      </div>
+    </div>
+  </div>
+  <div *ngSwitchCase="controlCategory.TEXT_BOX">
+
+  </div>
+  <div *ngSwitchCase="controlCategory.TEXT_AREA">
+    
+  </div>
```
Comment (rayliangatos) [inline src/hearings/components/multi-level-selector/multi-level-selector.component.html:25]: what are these two cases for? seems useless here could they be removed or you support these two types?

File: src/hearings/components/multi-level-selector/multi-level-selector.component.html [hunk 21]
```diff
@@ -0,0 +1,46 @@
...
+          <exui-multi-level-selector class="multi-container" *ngIf="childNode.value.child_nodes && childNode.value.child_nodes.length"
+            [configLevels]="configLevels"
+            [multiSelect]="getValue(childNode)"
+          [level]="level + 1">
+        </exui-multi-level-selector>
+        </div>
+    </div>
+  </div>
+  <div *ngSwitchCase="controlCategory.SELECT" [formGroup]="formGroup">
+    <label class="govuk-label" for="specialism1"> Choose specialism</label>
```
Comment (rayliangatos) [inline src/hearings/components/multi-level-selector/multi-level-selector.component.html:40]: no id matched 'specialism1'

Comment (rayliangatos) [reply src/hearings/components/multi-level-selector/multi-level-selector.component.html:40]: and a space before ' Choose specialism'

File: src/hearings/components/multi-level-selector/multi-level-selector.component.ts (hunk 17 above)
Comment (waynepHMCTS) [reply src/hearings/components/multi-level-selector/multi-level-selector.component.ts:11]: I should off been multLevelSelect. I will change to that instead.

**Review by rayliangatos — APPROVED** [review]

**Review by kuda1992 — APPROVED** [review]
Looks good. Thanks for making the changes @waynepHMCTS

**Pull request merged**